### ISSUE TITLE
WebCam remains active after track.disable()

### ISSUE THREAD

**Issue by GMatias93** [twilio/video-quickstart-js#48]
I slightly modified the media preview code a bit to pause/unpause tracks after they have been created. The screen is going black as expected, but the camera light remains on.

```javascript
// Preview LocalParticipant's Tracks.
document.getElementById('button-preview').onclick = function() {
  var localTracksPromise = previewTracks
    ? previewTracks.forEach(function(track) { track.enable(!track.isEnabled) })  //Toggle if exists
    : Video.createLocalTracks().then(function(tracks) {
    window.previewTracks = previewTracks = tracks;
    var previewContainer = document.getElementById('local-media');
    if (!previewContainer.querySelector('video')) {
      attachTracks(tracks, previewContainer);
    }
  }, function(error) {
    console.error('Unable to access local media', error);
    log('Unable to access Camera and Microphone');
  });
};
```

The desired behavior is for the user to toggle their videoTrack as they please (which is why `LocalVideoTrack.stop()` is not ideal), but because the user still sees the green light next to their webcam they understandably believe they are still streaming video. Is there another way I can force the webcam off while my videoTracks are disabled? Or will I have to resort to stopping tracks on `disable` and replacing it with a new one on `enable` (which I would then have to stitch together to make a single playback)?

Thank You!

**Comment by markandrus** [comment]
Hi @GMatias93,

[As of Firefox 60](https://blog.mozilla.org/webrtc/better-privacy-on-camera-mute-in-firefox-60/), Firefox will give you the webcam light behavior you desire when calling `disable`. This is because LocalTrack's `disable` method is really just setting the underlying MediaStreamTrack's `enabled` property to `false`, and Firefox will stop capturing when the `enabled` property is set to `false`.

 Unfortunately, not all browsers are created equally, and so, until and unless Chrome and Safari implement the same behavior, you'll need to `stop` & `unpublish` to "disable" and then call `createLocalTracks` (or `createLocalVideoTrack` and/or `createLocalAudioTrack`) & `publishTrack` to "enable".

**Comment by GMatias93** [comment]
@markandrus Thanks for getting back to me. We implemented a solution like the one suggested above 👍

**Comment by hmm29** [comment]
Hi @markandrus  and @GMatias93 ,

I'm working on implementing a solution like the one @markandrus suggested. I'm having a little bit of trouble getting my newly created local video track to be published to the room when "re-enabling" video. I'm using Twilio video SDK version 1.14.0.

In my code, I have a method called toggleLocalVideo that's supposed to handle the pause/unpausing of the user's local video track. Since enable/disable still leaves the webcam light on for Chrome users, I need to stop the localVideoTrack completely and then publish a new local video track to the room.

This is what I have so far.

```javascript
toggleLocalVideo = showLocalVideo => {
  const { localParticipant, room } = this.state;

  if (!localParticipant) return;

  if (showLocalVideo) {
      createLocalVideoTrack({width: 100}).then(track => {
          const localMediaContainer = document.getElementById('local-media-div');
          localMediaContainer.appendChild(track.attach());

         // THIS is the problem area: I want this new track to be published to the room 
         // so that when I disconnect, it will be removed
          room.localParticipant.addTrack(track);
      });
  } else {
    localParticipant.videoTracks.forEach((track) => {
      track.stop();
      const attachedElements = track.detach();
      attachedElements.forEach(element => element.remove());
    });
  }

  this.setState({ localVideoTrackEnabled: showLocalVideo });
};
```

also this is what my 'disconnected' event handler looks like:
```javascript
 room.on('disconnected', room => {
      localParticipantTracks.forEach(track => {
        const attachedElements = track.detach();
        attachedElements.forEach(element => element.remove());
      });

      remoteParticipants.forEach(participant => {
        participant.tracks.forEach(track => {
          const attachedElements = track.detach();
          attachedElements.forEach(element => element.remove());
        });
      });
    });
```
Any tips?

**Comment by manjeshbhargav** [comment]
Hi @hmm29 ,

You can use `LocalParticipant\'s` `publishTrack()` and `unpublishTrack()` in conjunction with `LocalVideoTrack`'s `stop()` like so:

```js
const toggleLocalVideo = showLocalVideo => {
  const { room: { localParticipant } } = this.state;
  if (showLocalVideo) {
    createLocalVideoTrack().then(track => {
      const localMediaContainer = document.getElementById('local-media-div');
      localMediaContainer.appendChild(track.attach());
      return localParticipant.publishTrack(track);
    });
  } else {
    const track = [...localParticipant.videoTracks.values()][0];
    track.stop();
    track.detach().forEach(element => element.remove());
    localParticipant.unpublishTrack(track);
  }
  this.setState({ localVideoTrackEnabled: showLocalVideo });
};
```

Let me know if this works for you.

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by hmm29** [comment]
Thank you @manjeshbhargav!!

**Comment by hmm29** [comment]
Hi @manjeshbhargav , one final follow-up question. With the above solution, whenever I re-publish my video track to the room, remote participants do not see my video track added back. They just see a black screen instead of seeing my video again. So, if I ever stop my local video track and then republish during a video chat, the remote participant won't be able to see my video again.

I'm wondering what I might be missing here. This is my event handler for remote participant connections. Inside it is the code to watch for a remote participant's 'trackAdded' events.

```javascript
// Log if Participant will connect to the room
    room.on('participantConnected', participant => {
      this.setState({ remoteParticipant: participant });

      console.log(`Participant "${participant.identity}" connected`);

      participant.tracks.forEach(track => {
        document.getElementById('remote-media-div').appendChild(track.attach());
      });

      participant.on('trackAdded', track => {
          document.getElementById('remote-media-div').appendChild(track.attach());
      });
    });
```
Could you let me know if I need to do something else in this code to acknowledge that a remote participant has republished a video or audio track?

Many thanks in advance!

**Comment by manjeshbhargav** [comment]
Hi @hmm29 ,

When you `stop` a `LocalVideoTrack`, it is no longer usable anymore. If you want to re-publish your video, you would have to create a new `LocalVideoTrack`. That is precisely the reason why in my code snippet I call `createLocalVideoTrack()` instead of re-using the existing stopped `LocalVideoTrack`.

Let me know if this helps. Please feel free to re-open if you are still experiencing this issue, or open a new one for a new issue.

Thank you for using Twilio Video.

Manjesh Malavalli
JSDK Team

**Comment by hmm29** [comment]
Hi @manjeshbhargav, I should have specified that I was doing exactly what you did in your snippet. It works as you said. I am not reusing the stopped video track but when publishing a new local video track created with createLocalVideoTrack, my remote participants don't see anything. 

I figured out what happened. This was simply due to the newly published track's video element being stacked under the stopped track's black screen in the remote participant's video chat window. Using prepend() instead of appendChild() for the newly created track's element fixed it for me. Thanks again for the help!!

**Comment by pedromiravaz** [comment]
Hi @manjeshbhargav , I tried to reproduce the code above to disconnect the camera:

`        gl_room.on('disconnected', room => {
            room.localParticipant.tracks.forEach(publication => {
                publication.track.stop();
                const attachedElements = publication.track.detach();
                attachedElements.forEach(element => element.remove());
            });
            const track = [...room.localParticipant.videoTracks.values()][0];
            track.stop();  // <- error
            track.detach().forEach(element => element.remove());
            room.localParticipant.unpublishTrack(track);
        });`

But on chrome I get a "track.stop is not a function" error, on line commented with <- error.Can you help?

**Comment by manjeshbhargav** [comment]
@pedromiravaz ,

In your code snippet, the section of code after `room.localParticipant.tracks.forEach(() => {...})` is not necessary since you have already stopped all the tracks. 

Having said that, the reason you are seeing the error is that `track` is actually a LocalTrackPublication. You can access the LocalTrack associated with it and stop it as follows:
```js
const videoTrack = [...room.localParticipant.videoTracks.values()][0].track;
videoTrack.stop();
```

As I said earlier, the last 4 lines in the snippet are unnecessary since you've already stopped all the tracks in `forEach`.

Please let me know if this works for you.

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by pedromiravaz** [comment]
Thank you @manjeshbhargav  for your prompt response!

Yes, it works now, but the camera still shows the light indicator showing it's being used. I want to disconnect and turn off the camera (that's why I tried the "unpublish").

´
        gl_room.on('disconnected', room => {
            room.localParticipant.tracks.forEach(publication => {
                publication.track.stop();
                const attachedElements = publication.track.detach();
                attachedElements.forEach(element => element.remove());

            });
        });

        const videoTrack = [...gl_room.localParticipant.videoTracks.values()][0].track;
        videoTrack.stop();
        gl_room.localParticipant.unpublishTrack(videoTrack);

        gl_room.disconnect();
´

How can I stop the camera?

**Comment by manjeshbhargav** [comment]
@pedromiravaz ,

Can you share your media capture logic, that is how you access the camera?

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by pedromiravaz** [comment]
Sure @manjeshbhargav  :)

```
        const Video = Twilio.Video;
        
        const { createLocalVideoTrack } = Twilio.Video;

        gl_localvideotrack = createLocalVideoTrack({
            audio: false,
            video: { width: 160 },
        }).then(track => {
            const localMediaContainer = document.getElementById('local-media');
            localMediaContainer.appendChild(track.attach());
        });

        Video.connect(tw_token_jwt, { name: tw_roomname, automaticSubscription: true }).then(room => {
            console.log(`Successfully joined a Room: ${room}`);
            gl_room=room;

            const localParticipant = room.localParticipant;
            console.log(`Connected to the Room as LocalParticipant "${localParticipant.identity}"`);

            room.participants.forEach(participant => {
                console.log(`Participant "${participant.identity}" is already connected to the Room`);
                participantConnected(participant);
            });
            room.once('participantConnected', participant => {
                console.log(`Participant "${participant.identity}" connected`);
                participantConnected(participant);
            });

        }, error => {
            console.error(`Unable to connect to Room: ${error.message}`);
        });

    });

    function trackPublished(publication, participant) {
        console.log(`RemoteParticipant ${participant.identity} published a RemoteTrack: ${publication}`);
      
        publication.on('subscribed', track => {
          console.log(`LocalParticipant subscribed to a RemoteTrack: ${track}`);
          document.getElementById('remote-media').appendChild(track.attach());
        });
      
        publication.on('unsubscribed', track => {
          console.log(`LocalParticipant unsubscribed from a RemoteTrack: ${track}`);
        });
    }

    function participantConnected(participant) {
        participant.tracks.forEach(publication => {
          trackPublished(publication, participant);
        });
      
        participant.on('trackPublished', publication => {
          trackPublished(publication, participant);
        });
      
        participant.on('trackUnpublished', publication => {
          console.log(`RemoteParticipant ${participant.identity} unpublished a RemoteTrack: ${publication}`);
        });
    }

```

**Comment by manjeshbhargav** [comment]
@pedromiravaz ,

I see what the problem is. Since you are not passing either media constraints or an array of pre-acquired tracks to [connect](), the sdk acquires audio and video tracks for you. Since you also call [createLocalVideoTrack](), there are now two camera tracks, one for your local preview and one for publishing to the Room. You are stopping the camera track published to the Room, but not the camera track used for your local preview.

The advantage of letting the sdk acquire media tracks for you is that it also manages their clean-up, that is stopping them once you leave the Room. So, you need not call `createLocalVideoTrack` for previewing your local video if you do this:
```js
Twilio.Video.connect(token, {
  audio: true,
  name: 'my-room',
  video: { /* video capture constraints */ },
}).then(room => {
  // Get the camera track for your preview.
  const cameraTrack = [...room.localParticipant.videoTracks.values()][0].track;
  localMediaContainer.appendChild(cameraTrack.attach());
});
```
Now, since the sdk was responsible for acquiring the media, it will also automatically stop it when you disconnect from the Room.

Please let me know if this works for you.

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by pedromiravaz** [comment]
Thanks a lot @manjeshbhargav ! Works beautifully now. I still need to further review your answer to understand better what's happening, but it's running like it should. Kudos to you.

**Comment by manjeshbhargav** [comment]
@pedromiravaz ,

I'm glad that I was able to help you out. Thanks for being a Twilio Video customer!

Manjesh Malavalli
JSDK Team

**Comment by RinaDesai** [comment]
Hello,
I also have same issue webcam light remain on when video call disconnect.

connect(token: string, roomName: string, callback: (err, res) => void) {
        connect(token, { name: roomName }).then(room => {
            this.activeRoom = room;
            this.roomHandler(room);
            callback(null, room);
        }, err => {
            callback(err, null);
        });
    }

    roomHandler(room) {
        room.participants.forEach(p => {
            this.participantConnected(p);
        })
        room.on('participantConnected', participant => {
            this.participantConnected(participant);
        });
        room.on('participantDisconnected', participant => {
            this.participantDisconnected(participant);
        });
        room.on('disconnected', function () {
            room.participants.forEach(function (participant) {
                participant.track.stop();
                const tracks = participant.track.detach();
                console.log("unsubscribed from: " + participant.track);
                tracks.forEach(element => element.remove());
                room.localParticipant.unpublishTrack(participant.track);
            });
            if (room.localParticipant) {
                room.localParticipant.tracks.forEach(participant => {
                    participant.track.stop();
                    const tracks = participant.track.detach();
                    console.log("unsubscribed from: " + participant.track);
                    tracks.forEach(element => element.remove());
                    room.localParticipant.unpublishTrack(participant.track);
                });
            }
            room.disconnect();
        });
    }

**Comment by cahergil** [comment]
@manjeshbhargav in this code
```
else {
    const track = [...localParticipant.videoTracks.values()][0];
    track.stop();
    track.detach().forEach(element => element.remove());
    localParticipant.unpublishTrack(track);
  }
```

 and considering i am not dealing with enable/disable of camera, just disconnecting the video call, is necessary to alwas call localParticipant.unpublishTrack(track) after calling track.stop()

I aks it because one person in my teams the light of the camera is not off when disconnecting, and I am not unpublishing the local participant tracks, just stopping them.
Thanks in advanced!

**Comment by Kumark1008** [comment]
In angular we are using a  twilio js sdk,
in that after  from room disconnect camara is not getting off.
  this.meetingRoom.disconnect();
      publication.track.disable();
        publication.track.stop();
        publication.track.detach().forEach(element => element.remove());
        this.meetingRoom.localParticipant.unpublishTrack( publication.track);

**Comment by IshikaSW** [comment]
> Hi @hmm29 ,
> 
> When you `stop` a `LocalVideoTrack`, it is no longer usable anymore. If you want to re-publish your video, you would have to create a new `LocalVideoTrack`. That is precisely the reason why in my code snippet I call `createLocalVideoTrack()` instead of re-using the existing stopped `LocalVideoTrack`.
> 
> Let me know if this helps. Please feel free to re-open if you are still experiencing this issue, or open a new one for a new issue.
> 
> Thank you for using Twilio Video.
> 
> Manjesh Malavalli
> JSDK Team

Hi Manjesh , 
I have a different issue , i am using js sdk for live video . When i switch camera on browser first i unpublish and stop existing track and republish video by creating local video track but with this when we transcode the video we get multiple file i.e , if i started previewing from back cam and then switch to front and then back cam again , transcoding will generate me 3 video files with 2 for back cam and one for front cam . This does not happen in android or ios apps . Can you please help me with this .

**Comment by Kumark1008** [comment]
Yeah i tried this method,didn't works.
after disconnecting twilio video / stopping the track also camera is not getting off.

**Comment by gajahab90** [comment]
Related to camera turn off, I solved this problem this way:
  ```
getLocalMedia = () => {
        Video.createLocalVideoTrack().then(track => {
            this.setState({localTrack: track});
        });
    }
```
then at the end call "stop()"
`this.state.localTrack.stop();`

**Comment by babu99k** [comment]
Hi @GMatias93 where you able to resolve this issue. I am also facing the same issue you mentioned above

**Comment by kybishop** [comment]
This bug still persists for us in Safari v15

Connection code:
```typescript
this.room = await TwilioVideo.connect(accessToken, {
  audio: true,
  name: channelKey,
  video: { aspectRatio: 1 }
});
const cameraTrack = [...this.room.localParticipant.videoTracks.values()][0].track;
this.localContainer.appendChild(cameraTrack.attach());
```

Disconnect code:
```typescript
this.room.localParticipant.tracks.forEach(publication => {
  const track = publication.track;
  if (track.kind == 'video' || track.kind == 'audio') {
    track.detach().forEach(element => element.remove());
    track.stop();
  }
  publication.unpublish();
});
this.room.disconnect();
```

**Comment by ArturV93** [comment]
> In angular we are using a twilio js sdk, in that after disconnecting from room, camara is not getting off. this.room.disconnect(); publication.track.disable(); publication.track.stop(); publication.track.detach().forEach(element => element.remove()); this.room.localParticipant.unpublishTrack( publication.track);

Hey, I have the same issue with Angular. Did you able to solve this issue?
